src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala

```diff
@@ -374,10 +374,43 @@ trait MatchApproximation extends TreeAndTypeAnalysis with ScalaLogic with MatchT
         def tru                                               = True
       }
 
+
+      private def isIrrefutabilityProof(sym: Symbol): Boolean = {
+        sym.isMethod && sym.name == TermName("isEmpty") && {
+          //In some cases this check returns true while checking for
+          //`ConstantFalse` returns false, for example when compiling
+          //test/files/neg/t12240.scala
```

Comment:
Maybe you were comparing the types with `==`? On types that's reference equality, for type equality you'd have to use `=:=`. Anyway, the current code LGTM, so I'd remove the comment, it's a bit confusing.

Reply:
I think this is related to the separate vs joint compilation issue of constant types.  ConstantFalse is the foldable kind and `neg/t12240.scala` defines `def isEmpty: false` which, IIUC, is a (non-foldable) literal type in joint compilation and only a foldable constant type if it were to be used in a separate compilation.